crates/runtime/src/flight.rs

```diff
@@ -0,0 +1,173 @@
+use arrow::ipc::writer::{DictionaryTracker, IpcDataGenerator};
+use std::sync::Arc;
+
+use crate::datafusion::DataFusion;
+use arrow_flight::{FlightEndpoint, SchemaAsIpc};
+use datafusion::arrow::error::ArrowError;
+use datafusion::datasource::file_format::parquet::ParquetFormat;
+use datafusion::datasource::listing::{ListingOptions, ListingTableUrl};
+use futures::stream::BoxStream;
+use tonic::transport::Server;
+use tonic::{Request, Response, Status, Streaming};
+
+use arrow_flight::{
+    flight_service_server::FlightService, flight_service_server::FlightServiceServer, Action,
+    ActionType, Criteria, Empty, FlightData, FlightDescriptor, FlightInfo, HandshakeRequest,
+    HandshakeResponse, PutResult, SchemaResult, Ticket,
+};
+
+pub struct Service {
+    data_fusion: DataFusion,
+}
+
+#[tonic::async_trait]
+impl FlightService for Service {
+    type HandshakeStream = BoxStream<'static, Result<HandshakeResponse, Status>>;
+    type ListFlightsStream = BoxStream<'static, Result<FlightInfo, Status>>;
+    type DoGetStream = BoxStream<'static, Result<FlightData, Status>>;
+    type DoPutStream = BoxStream<'static, Result<PutResult, Status>>;
+    type DoActionStream = BoxStream<'static, Result<arrow_flight::Result, Status>>;
+    type ListActionsStream = BoxStream<'static, Result<ActionType, Status>>;
+    type DoExchangeStream = BoxStream<'static, Result<FlightData, Status>>;
+
+    async fn get_schema(
+        &self,
+        request: Request<FlightDescriptor>,
+    ) -> Result<Response<SchemaResult>, Status> {
+        let request = request.into_inner();
+
+        let listing_options = ListingOptions::new(Arc::new(ParquetFormat::default()));
+        let table_path = ListingTableUrl::parse(&request.path[0]).map_err(to_tonic_err)?;
```

Comment:
We shouldn't use `map_err` to return errors, see the style guide: https://github.com/spiceai/spiceai/blob/rust/docs/dev/style_guide.md#utilize-context-to-encapsulate-underlying-errors-into-module-specific-errors

Reply:
This code isn't actually returning rust errors, it's translating the error into the json error response. I couldn't get that to work with snafu and my assumption is that it isn't possible since its not an error type.